rfcs/assert_throws.md

```diff
@@ -0,0 +1,45 @@
+# `testharness.js`: Replace `assert_throws` and `promise_rejects` with more fine-grained APIs
+
+## Summary
+
+The expected value passed to `assert_throws` and `promise_rejects` APIs can be one of several different kinds of values.
+This makes it harder to correctly compare the actual thrown exception / rejection value.
+Remove these APIs, and replace them with more focused APIs.
+
+## Details
+
+The `assert_throws` and `promise_rejects` APIs verify the value that is thrown or passed as a rejection value.
+They support being called with three kinds of expected values:
+
+- a string or number representing a `DOMException` (e.g., `"INVALID_STATE_ERR"`, `"InvalidStateError"`, or `DOMException.INVALID_STATE_ERR`);
+- an ECMAScript-style `Error` object (e.g., `new TypeError()`);
+- a plain object (in practice, the exact same object should be thrown or used as the rejection value).
+
+The last two are handled the same way: only the "name" properties are compared between the expected and the actual value.
+Because of the conflation of the various cases, the existing APIs don't get any of them quite right, and this causes some browser bugs to be missed.
+For example, the following passes:
+
+```js
+assert_throws(new TypeError(), () => { throw new DOMException("TypeError") });
+```
+
+Six new APIs are proposed to handle each of the cases separately:
+
+- `promise_rejects_dom` and `assert_throws_dom`, called with a string representing a `DOMException` (as above);
+- `promise_rejects_js` and `assert_throws_js`, called with the Error constructor (e.g. `TypeError`);
+- `assert_throws_exactly` and `promise_rejects_exactly`, called with the expected object and using `Object.is`.
+
+The new APIs will be added first, in order to allow a well-organized transition.
+Once this is done and browser vendors have updated their copies of `testharness.js`, we can start working on a (mostly automated) conversion of the tests in wpt.
+The old API will only be removed once all callers have been removed from wpt and browser vendors have had a chance to update any callers in proprietary tests.
```

Comment:
Is anyone signed up to do this work in each engine, or is this an ask for us to do it? Or would you do some engines but not all?

Looks like we have about 600 tests using `assert_throws` and `promise_rejects` in Chromium, about 300 of each.

Reply:
Not yet. I think it'd be best if the browser vendors handled this themselves; we can see how to deal if that doesn't happen.

Reply:
If the timeline is quite lax I think we can agree to handle it on the Chromium side. Can you get a similar commitment for the other repos so that this doesn't get stuck in a half-transitioned state for months and months?

Reply:
@jdm for Servo
@jgraham for Gecko
@youennf for WebKit (if you have proprietary tests that use upstream th.js at all)

Would y'all be willing to deal with this? I imagine we could set a deadline about two months after the RFC and https://github.com/web-platform-tests/wpt/pull/19054 land.

Reply:
I just checked, and if I am not misreading things Gecko has 2 uses of `assert_throws` and 0 uses of promise_rejects in our non-upstreamed tests.  I think we can commit to fixing that very expeditiously.  ;)

Reply:
Weird. @jgraham is it really the case that there's so little internal use of testharness.js in Gecko? What do people do with tests for a spec that they don't want to upstream yet but mean to in the future?

Reply:
Yes, I got the same results as @bzbarsky

I think this is a combination of two things:
* Gecko does a lot less work on speculative features
* We've been less successful at getting people to write wpt tests instead of gecko-specific tests unless they're specifically trying to contribute to wpt.

Reply:
>  I think we can commit to fixing that very expeditiously

https://bugzilla.mozilla.org/show_bug.cgi?id=1607307 tracks that.

Reply:
That was fast indeed! :)

Reply:
Well, terrible given the actual amount of work involved, but.... ;)